### PR TITLE
Refactor Publisher

### PR DESCRIPTION
**Pull request by daymxn**
Per [b/277637432](https://b.corp.google.com/issues/277637432),

This removes the `Publisher` class, and refactors the functionality that we still need into `BaseFirebaseLibraryPlugin`. This PR also comes with some optimizations and QOL improvements, such as;

- Moving our `afterEvaluate` to where it's actually needed, instead of wrapping all of our publishing configuration in it (which will allow earlier access to the publishing plugin)
- Usage of existing utility methods
- Significantly reduced overhead for accumulating aar artifacts during pom customization
- Additional general purpose util methods
- Links to new tracking bugs keeping an audit of potential sources of work (when features are added to tools, and bugs are fixed)

### CODE REVIEW

File: buildSrc/src/main/java/com/google/firebase/gradle/plugins/BaseFirebaseLibraryPlugin.kt [hunk 1]
```diff
@@ -94,22 +96,116 @@ abstract class BaseFirebaseLibraryPlugin : Plugin<Project> {
...
+ */
+// TODO(b/277607560): Remove when Gradle's MavenPublishPlugin adds functionality for aar types
+val FirebaseLibraryExtension.androidDependencies
+  get() =
+    project.configurations
+      .getByName(runtimeClasspath)
+      .incoming
+      .artifactView { attributes { attribute("artifactType", "aar") } }
+      .artifacts
+      .map { it.variant.displayName.substringBefore(" ") }
```
Comment (rlazo) [inline buildSrc/src/main/java/com/google/firebase/gradle/plugins/BaseFirebaseLibraryPlugin.kt:211]: related to the comment above

File: buildSrc/src/main/java/com/google/firebase/gradle/plugins/BaseFirebaseLibraryPlugin.kt [hunk 2]
```diff
@@ -94,22 +96,116 @@ abstract class BaseFirebaseLibraryPlugin : Plugin<Project> {
...
+   *   <type>aar</type>
+   * </dependency>
+   * ```
+   *
+   * @param dependency the element to append the `type` to
+   * @param androidLibraries a list of dependencies for this given SDK that publish `aar` artifacts
+   * @see applyPomTransformations
+   */
+  // TODO(b/277607560): Remove when Gradle's MavenPublishPlugin adds functionality for aar types
+  private fun addTypeWithAARSupport(dependency: Element, androidLibraries: List<String>) {
```
[the remark this answers is not in the document]
Comment (daymxn) [reply buildSrc/src/main/java/com/google/firebase/gradle/plugins/BaseFirebaseLibraryPlugin.kt:185]: This is in regards to the SDK's dependencies, not the project itself. So things such as androidx libs, play-services, etc.,

**Review by davidmotson — APPROVED** [review]

**Review by rlazo — APPROVED** [review]

File: buildSrc/src/main/java/com/google/firebase/gradle/plugins/publish/PublishingPlugin.java [hunk 3]
```diff
@@ -170,20 +171,22 @@ public void apply(Project project) {
               project
                   .getTasks()
                   .create(
```
Comment (rlazo) [inline buildSrc/src/main/java/com/google/firebase/gradle/plugins/publish/PublishingPlugin.java:173]: would register work?

File: buildSrc/src/test/kotlin/com/google/firebase/gradle/plugins/PublishingPluginTests.kt [hunk 4]
```diff
@@ -137,16 +138,21 @@ licenses {
...
       Project(name = "childProject2", version = "0.9", projectDependencies = setOf(project1))
 
     subprojectsDefined(project1, project2)
-    val exception =
-      Assert.assertThrows(UnexpectedBuildFailure::class.java) { publish(Mode.RELEASE, project2) }
-    assertThat(exception.message).contains("Failed to release com.example:childProject2")
+    val result = publishAndFail(RELEASE, project2)
+    assertThat(result.task(":checkHeadDependencies")?.outcome).isEqualTo(TaskOutcome.FAILED)
+    // also, if our integ tests fail, we need to add multidex back to installations and add a
+    // removal in our pom transformation for it
```
Comment (rlazo) [inline buildSrc/src/test/kotlin/com/google/firebase/gradle/plugins/PublishingPluginTests.kt:155]: wouldn't this be something we can test now and based on that either leave it out or put it back in?

Comment (daymxn) [reply buildSrc/src/test/kotlin/com/google/firebase/gradle/plugins/PublishingPluginTests.kt:155]: that comment was more for me, I talk to future me a lot in my comments :')

this pr is currently testing exactly that

File: buildSrc/src/test/kotlin/com/google/firebase/gradle/plugins/PublishingPluginTests.kt [hunk 5]
```diff
@@ -156,94 +162,95 @@ licenses {
...
-            artifactId = project1.name,
-            version = project1.latestReleasedVersion!!,
-            type = Type.AAR,
-            scope = "compile"
-          )
-        )
-      )
-  }
+  // @Test
+  // fun `Publish all dependent snapshot projects succeeds`() {
```
[the remark this answers is not in the document]
Comment (daymxn) [reply buildSrc/src/test/kotlin/com/google/firebase/gradle/plugins/PublishingPluginTests.kt:189]: If we remove the `snapshot` mode these tests would need to be removed. I haven't added support for it, and am waiting on any folks to respond to my chat about it before taking the overhead to add it. In the mean time, commenting out the tests. I'll have them either uncommented or completely removed before merging the PR.

File: buildSrc/src/main/java/com/google/firebase/gradle/plugins/publish/PublishingPlugin.java (hunk 3 above)
Comment (daymxn) [reply buildSrc/src/main/java/com/google/firebase/gradle/plugins/publish/PublishingPlugin.java:174]: Yes it would- changed :)

**Review by rlazo — APPROVED** [review]

**Pull request merged**